### DISCUSSION TITLE
Rate limit plugin

### DISCUSSION THREAD

**Discussion by yurirnascimento** [saloonphp/saloon#52] [General]
Hi,

One of my needs is to be able to use a request rate limit, because normally my jobs run for long periods but need to respect the API's request limit.

I'm creating a plugin to use the [Spatie/GuzzleRateLimiterMIddleware](https://packagist.org/packages/spatie/guzzle-rate-limiter-middleware)

Would it be useful to anyone else?

**Comment by Sammyjo20** [comment]
Yes, this would be great! I wonder if community made plugins could be released as standalone packages. Is this something you would be up for building?

**Comment by yurirnascimento** [comment]
This would be a useful plugin for me, so I could mantain it, what do you think?

**Comment by Sammyjo20** [comment]
If you are happy to, I would add it to a list of community driven plugins for Saloon.

**Comment by Sammyjo20** [comment]
Just a suggestion, it would be cool if the plugin also added two methods to "set" and "get" from a cache, so it would be really easy for Laravel users to hook it up to their Laravel app's cache

![Screen Shot 2022-01-28 at 21 42 13](https://user-images.githubusercontent.com/29132017/151625581-ff0d5e62-2af2-47e9-b47c-7926018571c3.png)

**Comment by yurirnascimento** [comment]
I'll probably do something simple, and expose the API for the user to customize as needed.

**Comment by Sammyjo20** [comment]
Sounds good. Looking forward to seeing it.

**Comment by Sammyjo20** [comment]
Just a thought, I need a rate limiter in Laravel, but I'd probably prefer if it threw an exception that I can catch in my job and delay it by X amount of seconds that the API replies with. Maybe that's something to consider in another plugin? Like a back-off plugin that uses Laravel Cache to determine if a rate limit has been exceeded

**Comment by yurirnascimento** [comment]
Hi, I'm still on it, I'll be back soon with news.

**Comment by Sammyjo20** [comment]
Sounds good @yurirnascimento no pressure on my side, take your time. Looking forward to seeing it.

**Comment by Sammyjo20** [comment]
I saw this online and thought it was really cool, perhaps we could use it for Saloon's rate limit plugin...

https://github.com/spatie/guzzle-rate-limiter-middleware

**Reply by yurirnascimento** [reply]
Hi Sammy, I did exactly with this Spatie plugin, I'm migrating an internal tool to use Saloon, sorry for the delay.

**Comment by Sammyjo20** [comment]
Hey @yurirnascimento 

Just wanted to let you know that I have been working on a new rate-limiting plugin for Saloon and I'm really excited about how it works. Would love to get your ideas/feedback on what I have built so far.

The goal with the new plugin is to act on prevention rather than reaction, so it's supposed to record the number of requests that you have made on a connector, and then if you have reached the number of requests within a given threshold, it will throw an exception or sleep/wait until the rate limit has lifted.

You first define the limits on your connector, as well as the rate limit store. The plugin will come with redis, predis, file, and PSR cache stores. You can have multiple limits too with lots of different intervals. 

```php
// In your connector

protected function resolveLimits(): array
{
    return [
         Limit::allow(100)->perMinute(),
         Limit::allow(1000)->perDay(),
         Limit::allow(10000)->perWeek()->name('custom-name'),
         Limit::fromTooManyRequests(retry: 60), // This one will stop all future requests if a response is 429
    ];
}
```

Next when you make requests, we will keep track of the number of hits per-limit and record it in the stores.

When it comes to Laravel queued jobs, you'll be able to check if there are any limits before making requests, for example at the top of your job, you could write something like:

```php
public function handle()
{
    if ($limit = $this->connector->getExceededLimit()) {
          $this->release($limit->getRemainingSeconds()); // Release job back onto queue to wait for the limit to be released
    }
}
```

Check out the source code here: https://github.com/Sammyjo20/saloon-rate-limiter-plugin

I'm still working on the tests and getting it working, but would love to hear if I've missed anything.
